bee-network/src/network/host.rs

```diff
@@ -0,0 +1,252 @@
+                .send(InternalEvent::AddressBound {
+                    address: multiaddr.clone(),
+                })
+                .expect("send error");
+
+            // INVESTIGATE
+            peerlist
+                .0
+                .write()
+                .await
+                .insert_local_addr(multiaddr)
+                .expect("insert_local_addr");
+        }
+        SwarmEvent::ConnectionEstablished { peer_id, .. } => {
+            debug!("Swarm event: connection established with {}.", alias!(peer_id));
+        }
+        SwarmEvent::ConnectionClosed { peer_id, .. } => {
+            debug!("Swarm event: connection closed with {}.", alias!(peer_id));
+        }
+        SwarmEvent::ListenerError { error } => {
+            error!("Swarm event: listener error {}.", error);
+        }
+        SwarmEvent::Dialing(peer_id) => {
+            // NB: strange, but this event is not actually fired when dialing. (open issue?)
+            debug!("Swarm event: dialing {}.", alias!(peer_id));
+        }
+        SwarmEvent::IncomingConnection { send_back_addr, .. } => {
+            debug!("Swarm event: being dialed from {}.", send_back_addr);
+        }
+        _ => {}
+    }
+}
+
+async fn process_command(command: Command, swarm: &mut Swarm<SwarmBehavior>, peerlist: &PeerList) {
+    match command {
+        Command::DialPeer { peer_id } => {
+            if let Err(e) = dial_peer(swarm, peer_id, peerlist).await {
+                warn!("Failed to dial peer '...{}'. Cause: {}", alias!(peer_id), e);
+            }
+        }
+        Command::DialAddress { address } => {
+            if let Err(e) = dial_addr(swarm, address.clone(), peerlist).await {
+                warn!("Failed to dial address '{}'. Cause: {}", address, e);
+            }
+        }
+        _ => {}
+    }
+}
+
+async fn dial_addr(swarm: &mut Swarm<SwarmBehavior>, addr: Multiaddr, peerlist: &PeerList) -> Result<(), Error> {
+    if let Err(e) = peerlist.0.read().await.allows_dialing_addr(&addr) {
+        warn!("Dialing address {} denied. Cause: {:?}", addr, e);
+        return Err(Error::DialingAddressDenied(addr));
+    }
+
+    info!("Dialing address: {}.", addr);
+
+    // FIXME
+    GOSSIP_ORIGIN.store(true, std::sync::atomic::Ordering::SeqCst);
+
+    Swarm::dial_addr(swarm, addr.clone()).map_err(|_| Error::DialingAddressFailed(addr))?;
+
+    Ok(())
+}
+
+async fn dial_peer(swarm: &mut Swarm<SwarmBehavior>, peer_id: PeerId, peerlist: &PeerList) -> Result<(), Error> {
+    if let Err(e) = peerlist.0.read().await.allows_dialing_peer(&peer_id) {
+        warn!("Dialing peer {} denied. Cause: {:?}", alias!(peer_id), e);
+        return Err(Error::DialingPeerDenied(peer_id));
+    }
+
+    // Unwrap:
+    // We just checked, that the peer is fine to be dialed.
+    let PeerInfo { address, alias, .. } = peerlist.0.read().await.info(&peer_id).unwrap();
+
+    info!("Dialing peer: {} ({}).", alias, alias!(peer_id));
+
+    // FIXME
```

Comment:
what does FIXME mean? maybe add more context to that too

Reply:
FIXME means that the current design is not ideal, and needs fixing at some point. It is of minor relevance though as it can only cause an issue with very low probability, so that other matters are higher on the priority list. The worst thing that can happen is that a peer doesn't get connected in the current attempt. But since there are many peers and many attempts at least one of them will finally succeed. The perfect impl. of this crate obviously has this issue fixed.